### PR TITLE
arch: arm: aarch32: add IRQ relay mechanism to ARMv7/8-M

### PR DESCRIPTION
**Pull request by e-rk**
This commit allows the `SW_VECTOR_RELAY` and `SW_VECTOR_RELAY_CLIENT` pair to be
enabled on the ARMv7-M architectures and covers
all additional interrupt vectors.

~~blocked by https://github.com/zephyrproject-rtos/zephyr/pull/26289~~

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

### CODE REVIEW

File: arch/arm/core/aarch32/irq_relay.S
Comment (e-rk) [inline arch/arm/core/aarch32/irq_relay.S:108]: Since not all 240 external interrupts are used on all MCUs, I'm wondering if it is possible to reduce the number of defined vectors to conserve flash memory.

Comment (nvlsianpu) [reply arch/arm/core/aarch32/irq_relay.S:108]: @anangl Is interrupts number available via DT somehow?

**Review by ioannisg — COMMENTED** [review]
Looks fine to me.
You need to change the commit message to reflect that this is also done for ARMv8-M architecture.

File: arch/arm/core/aarch32/irq_relay.S
Comment (ioannisg) [reply arch/arm/core/aarch32/irq_relay.S:108]: Yes, they are.
@nvlsianpu I am wondering: why do we need to define the full-vector if we are not building for a bootloader, but for a chain-loadable firwmare? That firmware will not use this vector, right? (Well, only the MSP and Reset words are needed).
This dates back to #4536, but can't find the reasoning.

Comment (anangl) [reply arch/arm/core/aarch32/irq_relay.S:108]: > @anangl Is interrupts number available via DT somehow?

There is `NUM_IRQS` in Kconfig. It could be used here.

Comment (e-rk) [reply arch/arm/core/aarch32/irq_relay.S]: `NUM_IRQS` helped here. Thank you @anangl  for suggestion.

Comment (e-rk) [reply arch/arm/core/aarch32/irq_relay.S]: @ioannisg Indeed, the last binary in the chain does not need the relay vector table. The `relocate_vector_table` in `aarch32/prep_c.c` would have to do nothing in this case.

Comment (nvlsianpu) [reply arch/arm/core/aarch32/irq_relay.S]: @e-rk  looks like chain-loaded application has build in __vector_relay_handler referenced in __vector_relay_table build which is not used. Am I right?

Comment (ioannisg) [reply arch/arm/core/aarch32/irq_relay.S]: no @e-rk, I don't think this is correct. The chained binary still needs to set the vector start on the  ram address (normally on 0x20000000), for the bootloader to forward interrupts. But it does not need the relay table.

Comment (e-rk) [reply arch/arm/core/aarch32/irq_relay.S]: Sorry I phrased it badly. I meant that the last binary *should* not need the vector relay handler, but that's not the case as it is now. Should I address this in this PR?

Comment (ioannisg) [reply arch/arm/core/aarch32/irq_relay.S]: No need, that's only an enhancement to save some space, but might require Kconfig API changes, so I suggest to ignore for now and get this PR fix only the extension of SW IRQ forwarding in Mainline Cortex-M

File: arch/arm/core/aarch32/prep_c.c
Comment (ioannisg) [inline arch/arm/core/aarch32/prep_c.c:40]: This is not good. You cannot skip VTOR initialization on a device that has VTOR. Even if you build a Bootloader, you can't guarantee that this is the first image in boot sequence, so VTOR needs to be determined.

Comment (ioannisg) [reply arch/arm/core/aarch32/prep_c.c:40]: Also, if this is required, particularly, if the relay handler is not at the beginning of flash. There are SoCs that require special sections before the ROM vector table, so VTOR requires to be set.

**Review by ioannisg — CHANGES_REQUESTED** [review]
We need to rework the logic around VTOR initialization.

**Review by ioannisg — COMMENTED** [review]
I think this could be a good extension for the SW Interrupt forwarding mechanism. Might have interesting use-cases beyond the Cortex-M Baseline and into the "Mainline" ecosystem.

I'd like a feedback by @d3zd3z on this, possibly also other ARM maintainers (@MaureenHelm , @galak fYI, any objections? I guess we can sandbox this with proper Kconfigs, so default logic does not change.)

File: arch/arm/core/aarch32/prep_c.c
Comment (nvlsianpu) [reply arch/arm/core/aarch32/prep_c.c:40]: so we want that only bootloader touch VTOR.

Comment (nvlsianpu) [reply arch/arm/core/aarch32/prep_c.c:40]: #ifdef CONFIG_CPU_CORTEX_M_HAS_VTOR

#ifdef CONFIG_XIP
#define VECTOR_ADDRESS ((uintptr_t)_vector_start)
#else
#define VECTOR_ADDRESS CONFIG_SRAM_BASE_ADDRESS
#endif

#else

#if defined(CONFIG_SW_VECTOR_RELAY)
Z_GENERIC_SECTION(.vt_pointer_section) void *_vector_table_pointer;
#endif

#define VECTOR_ADDRESS 0
#endif /* CONFIG_CPU_CORTEX_M_HAS_VTOR */



``` C
#if CONFIG_CPU_CORTEX_M_HAS_VTOR
static inline void relocate_vector_table(void)
#else
void __weak relocate_vector_table(void)
#endif /* CONFIG_CPU_CORTEX_M_HAS_VTOR */
{
#if defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) && !defined(CONFIG_BOOTLOADER_MCUBOOT)
	SCB->VTOR = VECTOR_ADDRESS & SCB_VTOR_TBLOFF_Msk;
	__DSB();
	__ISB();
#endif
#if !defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) || defined(CONFIG_SW_VECTOR_RELAY)
#if defined(CONFIG_XIP) && (CONFIG_FLASH_BASE_ADDRESS != 0) || \
    !defined(CONFIG_XIP) && (CONFIG_SRAM_BASE_ADDRESS != 0)
	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;
	(void)memcpy(VECTOR_ADDRESS, _vector_start, vector_size);
#elif defined(CONFIG_SW_VECTOR_RELAY)
	_vector_table_pointer = _vector_start;
#endif
#endif
}
```
	__DSB();
	__ISB();

#if defined(CONFIG_XIP) && (CONFIG_FLASH_BASE_ADDRESS != 0) || \
    !defined(CONFIG_XIP) && (CONFIG_SRAM_BASE_ADDRESS != 0)
	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;
	(void)memcpy(VECTOR_ADDRESS, _vector_start, vector_size);
#elif defined(CONFIG_SW_VECTOR_RELAY)
	_vector_table_pointer = _vector_start;
#endif
}

Comment (e-rk) [reply arch/arm/core/aarch32/prep_c.c]: I have applied the suggestion with one major change. The relay table address is under `__vector_relay_table` variable, therefore `VECTOR_ADDRESS` was set accordingly.

**Comment by nvlsianpu** [comment]
> I rebased and adapted to mainline changes.
I made the backup of @e-rk previous branch commit history here: https://github.com/nvlsianpu/zephyr/tree/backup/allow-irq-relay-for-armv7-m

**Comment by nvlsianpu** [comment]
@ioannisg Please revisit.

**Comment by ioannisg** [comment]
> @ioannisg Please revisit.

Yeap, will do this tomorrow morning.

File: arch/arm/core/aarch32/cortex_m/Kconfig
Comment (stephanosio) [inline arch/arm/core/aarch32/cortex_m/Kconfig:277]:
```suggestion
	  Normally this is only required for Cortex-M0(+) and Armv8-M baseline
	  cores with no hardware vector table relocation mechanisms (e.g. VTOR).
```

Comment (e-rk) [reply arch/arm/core/aarch32/cortex_m/Kconfig]: Done.

File: arch/arm/core/aarch32/irq_relay.S
Comment (ioannisg) [inline arch/arm/core/aarch32/irq_relay.S:29]: Is it possible to completely remove this #ifdef and make this conditional in CMakeLists?

Comment (ioannisg) [reply arch/arm/core/aarch32/irq_relay.S:29]: actually, i see that irq_relay.S is already compiled-in only when SW_VECTOR_RELAY is set

File: arch/arm/core/aarch32/cortex_m/Kconfig
Comment (ioannisg) [inline arch/arm/core/aarch32/cortex_m/Kconfig:274]: Just write "Cortex-M Baseline CPUs", instead of "Cortex-M0(+) and Armv8-M baseline cores"

File: arch/arm/core/aarch32/cortex_m/Kconfig
Comment (ioannisg) [inline arch/arm/core/aarch32/cortex_m/Kconfig:276]: Last sentence is not really needed. You can add a "not limited to" in the above sentence, after "is only required"

File: arch/arm/core/aarch32/irq_relay.S
Comment (e-rk) [reply arch/arm/core/aarch32/irq_relay.S:29]: Done.

Comment (nvlsianpu) [reply arch/arm/core/aarch32/irq_relay.S]: sounds reasonably ^^

File: arch/arm/core/aarch32/cortex_m/Kconfig
Comment (e-rk) [reply arch/arm/core/aarch32/cortex_m/Kconfig]: Done.

**Comment by nvlsianpu** [comment]
Tested on nRF51 and nRF52. Works fine.

**Review by nvlsianpu — APPROVED** [review]

**Review by ioannisg — COMMENTED** [review]
@e-rk could you please, explain why we need to do sw-vector-relay-client if platform has VTOR?
(i understand the use case with sw-vector-relay, i.e. a bootloader that wants to use sw vector relocation)

File: arch/arm/core/aarch32/prep_c.c
Comment (ioannisg) [inline arch/arm/core/aarch32/prep_c.c:70]: It's a bit difficult to see the use-case here: if I have VTOR, why do I need to be a chain-loadable leaf firmware client? I can just make VTOR point to my vector table.

Comment (ioannisg) [reply arch/arm/core/aarch32/prep_c.c:70]: To put it differently: in line 46 you set VECTOR_ADDRESS for the case where HAS_VTOR and RELAY_CLIENT are both set, but this value is not used anywhere.

**Comment by ioannisg** [comment]
Coming back to the original comment: If VTOR exists, it must always be set by the firmware during boot. This still allows you to do SW_VECTOR_RELAY as a bootloader, for non Mainline Cortex-M.

File: arch/arm/core/aarch32/prep_c.c
Comment (d3zd3z) [inline arch/arm/core/aarch32/prep_c.c:71]: I realize this was just moved from the code above, but I don't see how this would work. If the VECTOR_ADDRESS is not aligned according to the mask, it just won't work. Masking off the low bits won't help that (in fact, the hardware will happily do that for you when you write the value).

File: arch/arm/core/aarch32/prep_c.c
Comment (d3zd3z) [reply arch/arm/core/aarch32/prep_c.c:70]: I agree with wondering what the reason is to be able to have both VTOR and the relay feature enabled?

**Review by d3zd3z — COMMENTED** [review]
Have you been able to test the configuration with no VTOR (or is that why this can be enabled even without that), and the relay table, with the real vtable in mcuboot, relaying through a table setup by an app? If so, I'll add my approval.

**Comment by e-rk** [comment]
@ioannisg Good point. If the Zephyr binary it the only application (aside from the bootloader) that runs on a CPU with VTOR, there is no reason for it to be the sw relay client. It could be used however in legacy SW architectures where the Zephyr application is only a part of a bigger legacy system, however that would be a very rare use case I think. Should I block the CPUs with VTOR from enabling SW_RELAY_CLIENT?

File: arch/arm/core/aarch32/prep_c.c
Comment (e-rk) [reply arch/arm/core/aarch32/prep_c.c:71]: I agree, at least on Cortex-M4, the hardware does it for you.
From Cortex-M4 Devices Generic User Guide
>Table alignment requirements mean that bits[6:0] of the table offset are always zero.

**Comment by e-rk** [comment]
@d3zd3z After discussion with @ioannisg and @nvlsianpu, we decided to keep changes in `prep_c.c` to a minimum.
The flow would be as follows:
1. The mcuboot starts
1. Set the VTOR to the actual vector table of the mcuboot in `relocate_vector_table`
2. Do mcuboot processing
3. Before chain-loading the application, the VTOR is set to point to the relay vector table
4. If `CONFIG_BOOT_INTR_VEC_RELOC` is set, `_vector_table_pointer` to point to applications vector table
5. Chain-loaded Zephyr application always relocates the vector using the VTOR, if it is available. Otherwise it sets the `_vector_table_pointer`. It all happens in `relocate_vector_table`.

I have tested it on nRF51 and nRF52.
I'll make the appropriate PR to mcuboot later.

File: arch/arm/core/aarch32/cortex_m/relay_vector_table.ld
Comment (e-rk) [inline arch/arm/core/aarch32/cortex_m/relay_vector_table.ld:10]: I have removed this section and put the `__vector_relay_handler` in `.text`. That way, the two vector tables are one after another and it is more alignment-friendly.

I still have one unresolved problem however, and I'm not sure how to approach this.

The relay vector table gets put at the beginning, so it is always aligned. That is not the case for the real vector table of the application. The alignment is dependent on the number of interrupt handlers.

From Cortex -M4 Devices Generic User Guide

>When setting TBLOFF, you must align the offset to the number of exception entries in the vector
table. The minimum alignment is 32 words, enough for up to 16 interrupts. For more interrupts,
adjust the alignment by rounding up to the next power of two. For example, if you require 21
interrupts, the alignment must be on a 64-word boundary because the required table size is 37
words, and the next power of two is 64.

In `vector_table.ld` how would I have the `exc_vector_table` section be aligned to the nearest power of two depending on its size?

File: arch/arm/core/aarch32/cortex_m/Kconfig
Comment (ioannisg) [inline arch/arm/core/aarch32/cortex_m/Kconfig:279]: nit: move depends on below default (so the diff disappears)

File: arch/arm/core/aarch32/cortex_m/Kconfig
Comment (ioannisg) [inline arch/arm/core/aarch32/cortex_m/Kconfig:280]: Since you are here: maybe reword the help text, as follows:

```
Another image has enabled SW_VECTOR_RELAY, and will be forwarding
exceptions and HW interrupts to this image. Enable this option to make
sure the vector table pointer in RAM is set properly by the image upon
initialization.
```
@nvlsianpu agree this is more clear?

File: arch/arm/core/aarch32/cortex_m/vector_table.S
Comment (ioannisg) [inline arch/arm/core/aarch32/cortex_m/vector_table.S:86]: can you revert this change ?

File: include/linker/linker-defs.h
Comment (ioannisg) [inline include/linker/linker-defs.h:166]: Is this now needed @e-rk ?

File: arch/arm/core/aarch32/prep_c.c
Comment (d3zd3z) [reply arch/arm/core/aarch32/prep_c.c]: It isn't as much a matter of whether the mask is there or not, but that if the masked out bits are set, there isn't any hope of the vector table working. The HW doesn't add the VTOR value to the given address, it just uses the upper bits from the mask, and uses the lower bits from the vector. In other words, it doesn't matter whether you mask it or not.

The alignment depends on the particular SoC, based on the number of IRQ vectors used in that device.

It might be nice, at compile time, to detect the table being not at an aligned address.

**Review by d3zd3z — APPROVED** [review]

File: arch/arm/core/aarch32/cortex_m/relay_vector_table.ld
Comment (ioannisg) [reply arch/arm/core/aarch32/cortex_m/relay_vector_table.ld:10]: @e-rk can you file an issue about the vector table not being aligned?
I think this is a generic issue that we need to address, regardless of whether we do interrupt forwarding or not.

File: include/linker/linker-defs.h
Comment (e-rk) [reply include/linker/linker-defs.h:166]: This address will be used in mcuboot main.c

Comment (ioannisg) [reply include/linker/linker-defs.h:166]: ok thanks

File: arch/arm/core/aarch32/cortex_m/relay_vector_table.ld
Comment (ioannisg) [reply arch/arm/core/aarch32/cortex_m/relay_vector_table.ld:10]: Nevermind, I posted a PR that takes care of this (needs review, first but you can try it and see it if solves your problem)
https://github.com/zephyrproject-rtos/zephyr/pull/26800

Comment (e-rk) [reply arch/arm/core/aarch32/cortex_m/relay_vector_table.ld:10]: Thank you for doing this. It solves my problem. I'll remember now that Kconfig defines can be also used in ld scripts.

File: arch/arm/core/aarch32/cortex_m/Kconfig
Comment (e-rk) [reply arch/arm/core/aarch32/cortex_m/Kconfig:280]: I have applied the change.

File: arch/arm/core/aarch32/prep_c.c
Comment (ioannisg) [reply arch/arm/core/aarch32/prep_c.c]: >The alignment depends on the particular SoC, based on the number of IRQ vectors used in that device.

>It might be nice, at compile time, to detect the table being not at an aligned address.

@d3zd3z this is addressed now in #26800 . In fact this was missing in the code. Now we ensure at compile-time the vector table is properly aligned.

**Review by ioannisg — COMMENTED** [review]
This looks good to me now, @nvlsianpu @e-rk.

In brief, the PR allows a bootloader to use the SW_VECTOR_RELAY feature on Mainline Cortex-M variants.

Should we hold on and merge this after we resolve #26800?
@e-rk will you send up the MCUboot PR as well?

**Comment by e-rk** [comment]
@ioannisg Yes, let's wait for #26800.
I'll send the PR to MCUboot tomorrow, as I'm on the move today.

**Comment by nvlsianpu** [comment]
^^ sounds good

**Review by ioannisg — APPROVED** [review]

**Pull request merged**
